### ISSUE TITLE
Autolink and footnote extensions incorrectly process footnotes with certain letters

### ISSUE THREAD

**Issue by digitalmoksha** [github/cmark-gfm#121]
Ran across a strange interaction.  With footnotes and autolink enabled, certain footnote names, like with a `w` or an underscore in it, will cause the footnote not to be processed.  Here's an example:

```
The problem seems to be when there's a w or an _ in the footnote anchor.[^w] and [^a] and [^a_a]

[^w]: This won't be rendered.

[^a]: This will be rendered.

[^a_a]: This won't be rendered.
```

using `cmark-gfm --extension footnotes --extension autolink` gives

```
<p>The problem seems to be when there's a w or an _ in the footnote anchor.[^w] and <sup class="footnote-ref"><a href="#fn1" id="fnref1">1</a></sup> and [^a_a]</p>
<section class="footnotes">
<ol>
<li id="fn1">
<p>This will be rendered. <a href="#fnref1" class="footnote-backref">↩</a></p>
</li>
</ol>
</section>
```

Using version `cmark-gfm 0.28.3.gfm.16`

**Comment by tonyg** [comment]
A quick look at the code shows that [this test](https://github.com/github/cmark/blob/4ec9b35421fa7aed9b67a4d1ed5c1894e6498f1e/src/inlines.c#L1132) is failing for the footnote refs that are not being recognised - that is, `->next->next` is non-`NULL`. The quick experiment of disabling the check didn't help very much, but showed a possible related/adjacent problem, which is that the span considered [here](https://github.com/github/cmark/blob/4ec9b35421fa7aed9b67a4d1ed5c1894e6498f1e/src/inlines.c#L1134) is shorter than one might hope, in the failing cases not extending to just before the close bracket, and *varies depending on whether autolinking is enabled*. Perhaps there's some hidden state interference here? Could checking for autolinks earlier/above interfere with tokenization at this point?

**Comment by tnkdwp** [comment]
cmark-gfm --extension footnotes --extension autolink

**Comment by digitalmoksha** [comment]
@tnkdwp  ???

**Comment by kivikakk** [comment]
We add the special chars so the autolink extension gets called when it sees a `w` (such that it can autolink URLs starting with `www.`) and `:` (backtracks and looks for `http:`, `https:`, `ftp:`). The autolinker wasn't designed with the current footnotes extension in mind, and so there's as yet no solution to this.

**Comment by digitalmoksha** [comment]
Thanks for the update @kivikakk.

Hmmm....

**Comment by wolftune** [comment]
incidentally, this doesn't happen with capital W, only with lowercase.

**Comment by phillmv** [comment]
While standardizing on `cmark-gfm` for notes of mine, I realized this bug was impacting me and so I took a look at what's going on.

I chose to focus on just understanding how the footnote references work (thanks for the pointer @tonyg!), and after doing some thinking I think I came to a reasonable solution which can be found here: https://github.com/github/cmark-gfm/pull/227

tldr, i think it's safe to ignore that nodes that the autolinker is adding to the parser's ast (which is why the `->next->next` check Tony pointed out uh 3 years ago is failing), and in my own local use this patch seems to do the trick.

Reviews / feedback welcome!
